File: healthcare-connector/src/main/java/com/healthcare/connector/service/AuthorizationService.java
package com.healthcare.connector.service;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.healthcare.connector.entity.AuthorizationRequest;
import com.healthcare.connector.fhir.FHIRService;
import com.healthcare.connector.repository.AuthorizationRepository;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
public class AuthorizationService {

    private final AuthorizationRepository repository;
    private final GeminiService geminiService;
    private final FHIRService fhirService;
    public AuthorizationService(AuthorizationRepository repository, GeminiService geminiService, FHIRService fhirService) {
        this.repository = repository;
        this.geminiService = geminiService;
        this.fhirService = fhirService;
    }

    // Provider submits Authorization Request
    public AuthorizationRequest createRequest(AuthorizationRequest request) throws JsonProcessingException {

        request.setStatus("Pending");

        // Generate FHIR JSON
        String fhirJson = fhirService.generateClaim(request);
        request.setFhirJson(fhirJson);

        // Gemini Review
        String review = geminiService.reviewRequest(request);
        request.setAiReview(review);

        return repository.save(request);
    }
    // Provider & Payer View All Requests
    public List<AuthorizationRequest> getAllRequests() {
        return repository.findAll();
    }

    // View Single Request
    public AuthorizationRequest getRequest(Long id) {

        return repository.findById(id)
                .orElseThrow(() -> new RuntimeException("Authorization Request Not Found"));
    }

    // Payer Approves Request
    public AuthorizationRequest approve(Long id) {

        AuthorizationRequest request = getRequest(id);

        request.setStatus("Approved");

        return repository.save(request);
    }

    // Payer Rejects Request
    public AuthorizationRequest reject(Long id) {

        AuthorizationRequest request = getRequest(id);

        request.setStatus("Rejected");

        return repository.save(request);
    }

    // Delete Request
    public void deleteRequest(Long id) {

        repository.deleteById(id);
    }
}

File: healthcare-ui/src/app/services/authorization.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { AuthorizationRequest } from '../models/authorization-request';

@Injectable({
  providedIn: 'root'
})
export class AuthorizationService {

  private apiUrl = 'http://localhost:8080/api/request';

  constructor(private http: HttpClient) { }

  createRequest(request: AuthorizationRequest): Observable<AuthorizationRequest> {

    return this.http.post<AuthorizationRequest>(this.apiUrl, request);

  }

  getAllRequests(): Observable<AuthorizationRequest[]> {

    return this.http.get<AuthorizationRequest[]>(this.apiUrl);

  }

  approve(id: number): Observable<AuthorizationRequest> {

    return this.http.put<AuthorizationRequest>(`${this.apiUrl}/${id}/approve`, {});

  }

  reject(id: number): Observable<AuthorizationRequest> {

    return this.http.put<AuthorizationRequest>(`${this.apiUrl}/${id}/reject`, {});

  }
  // reviewRequest(request: AuthorizationRequest): Observable<string> {

  //   return this.http.post(
  //     `${this.apiUrl}/review`,
  //     request,
  //     { responseType: 'text' }
  //   );

  //}
  reviewRequest(request: AuthorizationRequest) {
    return this.http.post<any>(
      'http://localhost:8080/api/copilot/review',
      request
    );
  }
  deleteRequest(id: number) {

    return this.http.delete(
      `${this.apiUrl}/${id}`,
      { responseType: 'text' }
    );

  }
  getFHIR(id: number) {

    return this.http.get(
      `${this.apiUrl}/${id}/fhir`,
      {
        responseType: 'text'
      }
    );

  }

  getReview(id: number) {

    return this.http.get(
      `${this.apiUrl}/${id}/review`,
      {
        responseType: 'text'
      }
    );

  }
}